libgalois/include/katana/GraphTopology.h

```diff
@@ -0,0 +1,1175 @@
+    // between searching out going topology or incoming topology. However, incoming
+    // topology will return a different range of incoming edges instead of outgoing
+    // edges. Can we convert easily between outing and incoming edge range
+    if (Base::out().degree(src) == 0 || Base::in().degree(dst) == 0) {
+      return MakeStandardRange<edge_iterator>(Edge{0}, Edge{0});
+    }
+
+    return Base::out().FindAllEdgesSingleType(src, dst);
+  }
+
+  /// Check if vertex src is connected to vertex dst with the given edge edge_type
+  ///
+  /// @param src source node of the edge
+  /// @param dst destination node of the edge
+  /// @param edge_type edge_type of the edge
+  /// @returns true iff the edge exists
+  bool IsConnectedWithEdgeType(
+      Node src, Node dst, const EntityType& edge_type) const {
+    const auto d_out = Base::out().degree(src, edge_type);
+    const auto d_in = Base::in().degree(dst, edge_type);
+    if (d_out == 0 || d_in == 0) {
+      return false;
+    }
+
+    if (d_out < d_in) {
+      return Base::out().IsConnectedWithEdgeType(src, dst, edge_type);
+    } else {
+      return Base::in().IsConnectedWithEdgeType(dst, src, edge_type);
+    }
+  }
+
+  /// Check if vertex src is connected to vertex dst with any edge edge_type
+  ///
+  /// @param src source node of the edge
+  /// @param dst destination node of the edge
+  /// @returns true iff the edge exists
+  bool IsConnected(Node src, Node dst) const {
+    const auto d_out = Base::out().degree(src);
+    const auto d_in = Base::in().degree(dst);
+    if (d_out == 0 || d_in == 0) {
+      return false;
+    }
+
+    if (d_out < d_in) {
+      return Base::out().IsConnected(src, dst);
+    } else {
+      return Base::in().IsConnected(dst, src);
+    }
+  }
+
+
+};
+
+template <typename Topo>
+class BasicPropGraphViewWrapper : public Topo {
+  using Base = Topo;
+
+public:
+  explicit BasicPropGraphViewWrapper(
+      const PropertyGraph* pg, const Topo& topo) noexcept
+      : Base(topo), prop_graph_(pg) {}
+
+  const PropertyGraph& property_graph() const noexcept { return *prop_graph_; }
+
+private:
+  const PropertyGraph* prop_graph_;
+};
+
+// TODO (amber): add template flags for checking sort and transpose status
+using PGViewEdgesSortedByDestID =
+    BasicPropGraphViewWrapper<EdgesSortedByDestTopology>;
+using PGViewNodesSortedByDegreeEdgesSortedByDestID =
+    BasicPropGraphViewWrapper<NodesSortedByDegreeEdgesSortedByDestIDTopology>;
+using PGViewBiDirectional = BasicPropGraphViewWrapper<SimpleBiDirTopology>;
+using PGViewEdgeTypeAwareBiDir =
+    BasicPropGraphViewWrapper<EdgeTypeAwareBiDirTopology>;
+
+struct PropertyGraphViews {
+  using BiDirectional = PGViewBiDirectional;
+  using EdgesSortedByDestID = PGViewEdgesSortedByDestID;
+  using EdgeTypeAwareBiDir = PGViewEdgeTypeAwareBiDir;
+  using NodesSortedByDegreeEdgesSortedByDestID =
+      PGViewNodesSortedByDegreeEdgesSortedByDestID;
+};
+
+template <typename PGView>
+struct PGViewBuilder {};
```

Comment:
I don't like the contraction `PG` for types in the `katana` namespace. Can we move these into `katana::internal` if they belong there or expand the contraction?

Reply:
Yes, we can.